src/wallet/bdk.rs

```diff
@@ -1363,6 +1451,142 @@ impl Wallet {
     }
 }
 
+//___________________WALLET/DESCRIPTOR______________________________________________________________
+
+/// Wrapper for `IntoWalletDescriptor` that performs additional checks on the keys contained in the
+/// descriptor
+pub(crate) fn into_wallet_descriptor_checked<T: IntoWalletDescriptor>(
+    inner: T,
+    secp: &Secp256k1<All>,
+    network: Network,
+) -> Result<(ExtendedDescriptor, KeyMap), DescriptorError> {
+    let (descriptor, keymap) = inner.into_wallet_descriptor(secp, network)?;
+
+    // Ensure the keys don't contain any hardened derivation steps or hardened wildcards
+    let descriptor_contains_hardened_steps = descriptor.for_any_key(|k| {
+        if let DescriptorPublicKey::XPub(DescriptorXKey {
+            derivation_path,
+            wildcard,
+            ..
+        }) = k
+        {
+            return *wildcard == Wildcard::Hardened
+                || derivation_path
+                    .into_iter()
+                    .any(bdk_wallet::bitcoin::bip32::ChildNumber::is_hardened);
+        }
+
+        false
+    });
+    if descriptor_contains_hardened_steps {
+        return Err(DescriptorError::HardenedDerivationXpub);
+    }
+
+    if descriptor.is_multipath() {
+        return Err(DescriptorError::MultiPath);
+    }
+
+    // Run miniscript's sanity check, which will look for duplicated keys and other potential
+    // issues
+    descriptor.sanity_check()?;
+
+    Ok((descriptor, keymap))
+}
+
+impl Wallet {
+    fn create_wallet_descriptors(&self) -> Result<HashMap<KeychainKind, String>, anyhow::Error> {
```

Comment:
> can we just return HashMap<KeychainKind, String> instead of just wrapping inside Result enum?
because in this function , we are using unwrap at all places.

We should not be using `unwraps` at all. We have to make specific error kinds that can capture the required errors. We already have a `ProtocolError` type. Unwraps here should be replaced with appropriate error propagation as `ProtocolError`. And return should always be Result<stuff, ProtocolError>`.

But this can be done as finishing touches. For now, unwraps are okay, and you can omit the Result. Also we should not be using `anyhow`.

Although doing the `Result<stuff, ProtocolError>` from the very beginning will make the error propagation work later easier.

Reply:
This is also a single use function, and you can place the code at call site. And try to do it via Xpubs.

There is a clear way to do. Just explore the APIs.   

`Seed Phrase -> Master Xpriv -> Keychain Xprivs -> KeyMap = BTreeMap<DescriptorPublicKey, DescriptorSecretKey>`

You can directly use the BDK `Keymap` struct as `DescriptorPublickey` and `DescriptorSecretkey` are wrappers over Xpubs and Xprivs. You already have that via keychain derivations. 

Not trying to do the descriptor string will make life so much simpler for fidelity, swapcoin and contract keychains.

Reply:
Regarding  error enums, IMO currently having so many such enums makes error handling complicated .
also these errors are not compatible with BDK error ones.
So I feel there should be 1 unified error enum for whole coinswap (at least for wallet, I guess).
one good example-> 

https://github.com/MutinyWallet/mutiny-node/blob/2df5362e93100aced230082ffa495d3a593d2eb4/mutiny-core/src/error.rs#L20

Reply:
Yes thats what we do. https://github.com/citadel-tech/coinswap/blob/master/src/wallet/error.rs

Errors here could be included in a `Wallet::BdkError` error variant. Not `ProtocolError`. Thats a one level deeper error.

Note that the WalletError can hold Protocol Error too.

>

Reply:
Assuming resolved.